chore: Updates to serialization docs for 1.7 (#1121)
Unity-Technologies/com.unity.multiplayer.docs: docs/basics/networkvariable.md

    - `void ReadDelta(FastBufferReader reader, bool keepDirtyDelta)`
- Depdending upon your custom `NetworkVariableBase` container, you might look at `NetworkVariable<T>` or `NetworkList` to see how those two examples were implemented.

#### NetworkVariableSerialization&lt;T&gt;

> When working with known, non-generic types, it is easy and efficient to read and write using `FastBufferReader.ReadValue` and `FastBufferWriter.WriteValue`, respectively, or for integer types, using `BytePacker` and `ByteUnpacker` to add compression. When working with generic types, however, we have to rely on serializers that are generated based on types discovered during a compile-time code generation process. As such, when working with generic types, you have to a little bit of extra work to tell our code generation algorithm which types to generate serializers for. You can do this in two ways:

Author: > - Integer types:  Use `BytePacker` and `ByteUnpacker` to compress the network variable value. 

I would make it clear that this is optional and is a trade-off between bandwidth and CPU usage.

> To tell Unity which types to serialize, use `GenerateSerializationForTypeAttribute` for hard-coded types, and `GenerateSerializationForGenericParameterAttribute` for generic types. The following sections give an example of how to use these methods.

This ONLY applies to generic types, so it should be indented to make sure it's included in the bullet point for generics. Other types do not need to use these attributes.

> When working with known, non-generic types, it is easy and efficient to read and write using `FastBufferReader.ReadValue` and `FastBufferWriter.WriteValue`, respectively, or for integer types, using `BytePacker` and `ByteUnpacker` to add compression. When working with generic types, however, we have to rely on serializers that are generated based on types discovered during a compile-time code generation process. As such, when working with generic types, you have to a little bit of extra work to tell our code generation algorithm which types to generate serializers for. You can do this in two ways:

Author: 1. BytePacker and ByteUnpacker add compression/decompression to integer types. This makes them smaller (saving bandwidth) but requires analysis, mathematical manipulation of the value, and writing to/reading from the buffer one byte at a time, which uses CPU time.
2. No, it means you have two options:

```csharp
// Not hard-coded, generates serialization for whatever T happens to be
[GenerateSerializationForGenericParameter(0)]
public class MyNetVar<T>: NetworkVariableBase
{

}

// Hard-coded, specifically generates for Foo and nothing else
[GenerateSerializationForType(typeof(Foo))]
public class MyNetVar : NetworkVariableBase
{
}
```

The first way is to use `GenerateSerializationForTypeAttribute` to generate serialization for a specific hard-coded type:
```csharp
[GenerateSerializationForType(typeof(Foo))]
public class MyNetworkVariableTypeUsingFoo : NetworkVariableBase {}
```

> However, this is rarely useful for generic types, since if you know the type, you can generally use the more efficient route of calling `FastBufferReader`/`FastBufferWriter` methods directly.

Author: That statement is poorly worded, sorry... This will answer your other question below, too.

In a generic type, you have `GenericType<T>` where T is a generic type parameter (an unknown type).

When you do `[GenerateSerializationForType(typeof(Foo))]` what that does is it allows you to call `NetworkVariableSerialization<Foo>.Write()` and `NetworkVariableSerialization<Foo>.Read()`.

But if you know the type you are using is `Foo`, as opposed to the generic type parameter `T` which is an unknown type, you can usually just call `FastBufferWriter.WriteValue()` and `FastBufferReader.ReadValue()`, which do not need that kind of codegen-based initialization and are generally more efficient.

But if you want to serialize that value of type `T`, you can't call methods on `FastBufferWriter` and `FastBufferReader` because `T` is an unknown type and those types can't handle that. `NetworkVariableSerialization<T>` is a higher-level concept that uses codegen to determine how to serialize unknown types like that, but the codegen has to know what types to generate for.

with `[GenerateSerializationForGenericParameter()]` the codegen does some magic to figure out what `T` is and generate serialization for it, which is useful.

`[GenerateSerializationForType()]` is for if, for some reason, you know the type, but you want to call `NetworkVariableSerialization<T>` methods on it instead of `FastBufferReader`/`FastBufferWriter` methods. In 1.8.0 there isn't a whole lot of value in that - I can't think of an actual use case, but that doesn't mean there isn't one, so I left this there just in case.

In 1.9.0 there will be methods added for doing delta serialization of the differences between a current value and a previous one, so this will be more useful then.

So with that in mind... it may be best to just remove the comments about the usefulness of these methods entirely.

> However, this is rarely useful for generic types, since if you know the type, you can generally use the more efficient route of calling `FastBufferReader`/`FastBufferWriter` methods directly.

Reviewer: This explination helped _a lot_ thank you!
I've made a small change to the succestion that hopefully makes this clearer, but if you'd still prefer to remove it that's okay.

The more common and more useful method is to use `GenerateSerializationForGenericParameterAttribute` to generate serialization for a specific generic parameter in your `NetworkVariable` type:
```csharp
[GenerateSerializationForGenericParameter(0)]
public class MyNetworkVariableType<T> : NetworkVariableBase {}
```

This attribute takes an integer indicating which parameter in the type to generate serialization for. This value is 0-indexed, so the first type is 0, the second is 1, and so on. The attribute can be placed more than once on one class to generate serialization for multiple types:

```csharp
[GenerateSerializationForGenericParameter(0)]
[GenerateSerializationForGenericParameter(1)]
public class MyNetworkVariableType<TFirstType, TSecondType> : NetworkVariableBase {}
```

> This will generate serialization for both `TFirstType` and `TSecondType`.

Author: I'm curious why this should be removed, since it is there to point out that doing it this way can generate for multiple types?

> Either of these attributes will cause the code generation to generate the following methods that you can use:

Author: yes

```csharp
NetworkVariableSerialization<T>.Write(FastBufferWriter writer, ref T value);
NetworkVariableSerialization<T>.Read(FastBufferWriter writer, ref T value);
NetworkVariableSerialization<T>.Duplicate(in T value, ref T duplicatedValue);
NetworkVariableSerialization<T>.AreEqual(in T a, in T b);
```

> In cases where the type is dynamically allocated (managed types and collections like NativeArray and NativeList), calling `Read` will read the value in-place whenever possible, avoiding further allocations.

Author: It means we don't create a new object, but re-use the old one and write the new data into it directly.

> In cases where the type is dynamically allocated (managed types and collections like NativeArray and NativeList), calling `Read` will read the value in-place whenever possible, avoiding further allocations.

Reviewer: What happens if the value can't be read in-place? Is it not read at all or is it read as `inplace=False` by default? Is there any way the user can tell if the value can't be read in-place?

> In cases where the type is dynamically allocated (managed types and collections like NativeArray and NativeList), calling `Read` will read the value in-place whenever possible, avoiding further allocations.

Author: The only case where a value can't be read in-place is if the type is a managed type and the current value is `null`. It can't be read in-place because there is no current "place" to read to, so it allocates a new one. After that, it reads in-place.

`AreEqual` can be used to determine if a value has changed from a previous value cached via `Duplicate` at a previous time to determine if a value has changed or not, to avoid re-sending the same value multiple times. This previous value cached via `Duplicate` can also be used to calculate deltas for implementing `ReadDelta` and `WriteDelta`.

> Note that the type must be serializable according to the "Supported Types" list above. Each type needs its own serializer instantiated, so this step tells the codegen which types to create serializers for.

Reviewer: ```suggestion
The type you use must be serializable according to the "Supported Types" list above. Each type needs its own serializer instantiated, so this step tells the codegen which types to create serializers for.
```

> Note also that the codegen makes the assumption that all `NetworkVariable` types will exist as fields within `NetworkBehaviour` types, and limits its inspection to those fields to identify which types to create serializers for.

Reviewer: What is "codegen" in this case? Is it just Unity's code generator / compiler?

Author: We have a step that runs after compiling using ILPP (IL Post-Processing) to inject additional automatically generated code into the assembly. That process scans the assembly for all types that inherit from NetworkBehaviour, inspects all the fields in those types to see if they are NetworkVariables, uses the attributes mentioned above to find out what kind of types those NetworkVariables will serialize, and then outputs code for each of the types it found to make them serializable based on various attributes of the type (is it a class or a struct, is it an INetworkSerializable type, is it an INetworkSerializeByMemcpy type, is it a C# built-in type, and so on)

Author: The user can interact with the process by using the `GenerateSerializationFor*` attributes, but otherwise, it is automatic and happens behind the scenes. I personally think that it is useful for a user to know that the process exists and what it does, but they do not need to know the details. Personally, when I am a user of something that feels like "magic", I want it to be documented what that magic is and does so I can understand how the system I'm using works, at least at a high level.

Reviewer: Thanks for explaining that :) Appreciate it a lot. I mostly needed to know what to call that process so I've gone with "code generator" in my suggestion.

> Note also that the codegen makes the assumption that all `NetworkVariable` types will exist as fields within `NetworkBehaviour` types, and limits its inspection to those fields to identify which types to create serializers for.

Reviewer: ```suggestion
:::note Unity's code generator assumes that all `NetworkVariable` types exist as fields inside `NetworkBehaviour` types. This means that Unity only inspects fields inside `NetworkBehaviour` types to identify the types to create serializers for.
```

```csharp
[GenerateSerializationForGenericParameter(0)]
public class MyNetworkVariableType<T> : NetworkVariableBase {}
```

This attribute accepts an integer that indicates which parameter in the type to generate serialization for. This value is 0-indexed, which means that the first type is 0, the second type is 1, and so on. 
The following code example places the attribute more than once on one class to generate serialization for multiple types, in this case,`TFirstType` and `TSecondType:

```csharp
[GenerateSerializationForGenericParameter(0)]
[GenerateSerializationForGenericParameter(1)]
public class MyNetworkVariableType<TFirstType, TSecondType> : NetworkVariableBase {}
```

The  `GenerateSerializationForGenericParameterAttribute` and `GenerateSerializationForTypeAttribute` attributes make Unity's code generation create the following methods:

```csharp
NetworkVariableSerialization<T>.Write(FastBufferWriter writer, ref T value);
NetworkVariableSerialization<T>.Read(FastBufferWriter writer, ref T value);
NetworkVariableSerialization<T>.Duplicate(in T value, ref T duplicatedValue);
NetworkVariableSerialization<T>.AreEqual(in T a, in T b);
```

> In cases where the type is dynamically allocated (managed types and collections like NativeArray and NativeList), calling `Read` will read the value in-place whenever possible, avoiding further allocations.

Reviewer: ```suggestion
For dynamically allocated types with a value that isn't `null` (for example, managed types and collections like NativeArray and NativeList) call `Read` to read the value in the existing object and write data into it directy (in-place). This avoids more allocations. 
```

Reviewer: Added a suggestion - thankyou!
